Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,332 @@
+      simpa only [map_add] using congr_arg₂ (· + ·) h h')
+
+private lemma tensorModSMul_mul_smul {X' : Type v'} [AddCommGroup X'] [Module R X']
+    (s s' : S)
+    (z : X' ⊗[R] X) :
+    (s * s') • z = s • s' • z :=
+  z.induction_on rfl
+    (fun x' x => by simp only [tensorModSMul_smul_tmul, mul_smul])
+    (fun _ _ (h : map _ _ _ = _) (h' : map _ _ _ = _) => by
+        show map _ _ (_ + _) = map _ _ (map _ _ _)
+        rw [map_add, h, h', map_add, map_add]
+        rfl)
+
+private lemma tensorModSMul_add_smul {X' : Type v'} [AddCommGroup X'] [Module R X']
+    (s s' : S)
+    (z : X' ⊗[R] X) :
+    (s + s') • z = s • z + s' • z :=
+  z.induction_on rfl
+    (fun x' x => by simp only [tensorModSMul_smul_tmul, mul_smul, add_smul, tmul_add])
+    (fun z z' (h : map _ _ _ = _) (h' : map _ _ _ = _) => by
+        show map _ _ (_ + _) = map _ _ _ + map _ _ _
+        rw [map_add, h, h', map_add, map_add]
+        change _ = (s • z + s • z') + (s' • z + s' • z')
+        abel)
+
+noncomputable local instance tensorMod {X' : Type v'} [AddCommGroup X'] [Module R X'] :
+    Module S (X' ⊗[R] X) where
+  smul := (· • ·)
+  one_smul := tensorModSMul_one_smul _ _ _
+  mul_smul := tensorModSMul_mul_smul _ _ _
+  smul_zero := fun _ => rfl
+  smul_add := fun _ _ _ => map_add _ _ _
+  add_smul := tensorModSMul_add_smul _ _ _
+  zero_smul := fun z => z.induction_on rfl
+    (fun _ _ => by rw [tensorModSMul_smul_tmul, zero_smul, tmul_zero])
+    (fun _ _ (h : map _ _ _ = _) (h' : map _ _ _ = _) => show map _ _ _ = _ by
+      rw [map_add, h, h', zero_add])
+
+/--
+Let `R` be a commutative ring and `S` a ring.
+Give `(S, R)`-bimodule `X`, a left `R`-module `X` and a right `S`-module `Y`, then the the set of
+`S`-linear maps from `X` to `Y` has a (left) `R`-module structure given by:
+```
+l : X →ₗ[Sᵐᵒᵖ] Y
+r : R
+x : X
+-------------
+(r • l) x = l (r • x)
+```
+-/
+local instance hom_bimodule {Y : Type v''} [AddCommGroup Y] [Module S Y] :
+    Module R (X →ₗ[S] Y) where
+  smul r l :=
+  { toFun := fun x => l (op r • x)
+    map_add' := fun x y => by dsimp; rw [smul_add, map_add]
+    map_smul' := fun s x => by dsimp; rw [smul_comm, l.map_smul] }
+  one_smul l := LinearMap.ext fun x => show l _ = _ by rw [op_one, one_smul]
+  mul_smul r₁ r₂ l := LinearMap.ext fun x => show l _ = l _ by rw [op_mul, mul_smul]
+  smul_zero r := rfl
+  smul_add r l₁ l₂ := LinearMap.ext fun x => show (l₁ + _) _ = _ by
+    rw [LinearMap.add_apply, LinearMap.add_apply]; rfl
+  add_smul r₁ r₂ l := LinearMap.ext fun x => show l _ = l _ + l _ by
+    rw [op_add, add_smul, map_add]
+  zero_smul l := LinearMap.ext fun x => show l _ = 0 by rw [op_zero, zero_smul, map_zero]
+
+variable {R S X}
+/--
+Let `R` be a commutative ring and `S` a ring.
+Give `(S, R)`-bimodule `X`, a left `R`-module `X` and a right `S`-module `Y`,
+any `S`-linear map `X ⊗[R] X' ⟶ Y` can by curried into a bilinear map `X' ⟶ X ⟶ Y` where
+the first argument is `R`-linear and second is `S`-linear.
```

Comment:
```suggestion
Let `R` be a commutative ring and `S` a ring.
Given an `(S, R)`-bimodule `X`, a left `R`-module `Y` and a left `S`-module `Z`,
any `S`-linear map `X ⊗[R] Y ⟶ Z` can by curried into a bilinear map `Y ⟶ X ⟶ Z` that
is `R`-linear in the first argument and `S`-linear in the second.
```
maybe we can preemptively remove the "commutative" too, even though we haven't generalized tensor products yet ...